### DISCUSSION TITLE
Toggle lists & cursor type directly in code blocks

### DISCUSSION THREAD

**Discussion by brainbow1** [zim-desktop-wiki/zim-desktop-wiki#1667] [Ideas]
Two feature ideas that I want to contribute:

a) Toggle Lists (as in Notion)

b) When I insert a code-block via my preferred keybinding, there pops up that window "insert code block", where u can choose the preferred syntax/language, but this cannot be pushed by pressing Enter, I have to click OK.
After clicking OK, there appears a codeblock. But I can't immediately start typing in it but have to click in it first, to position the cursor in it.
Would improve efficiency when I can just confirm insertig the codeblock with Enter, confirm the last used language (or choose another one), and just start typing immediately inside the codeblock.

That may be luxus problems/ideas, but I would find it an amazing future.. don't know how hard it would be to realize something like that.

**Comment by introt** [comment]
I'd love Toggle Lists for active recall. Currently using headers, ToC and careful glances inside Zim, but I've been meaning to make a HTML template for this (edit: [here](https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1667#discussioncomment-1749706)) (no need to export after every change btw, just start the integrated webserver and hit refresh in the browser). It's quite simple with some [JS](https://www.w3schools.com/howto/howto_js_close_list_items.asp), and can be done with pure HTML as well using the `<summary>` element.

**Reply by introt** [reply]
As for implementing toggle lists inside Zim, I don't see why it wouldn't be possible right now, though waiting for [embedded pageviews](https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/1590#issuecomment-922453163) might make it possible to eg. integrate parts of lists from other pages

**Reply by introt** [reply]
Whereas b) is a lot simpler, "just" need to store & retrieve the last language selection wherever state goes && point the focus at the right stuff

**Comment by jaap-karssenberg** [comment]
Can you explain what "toggle list" are and for what use case those can be
useful ?

-- Jaap

**Reply by brainbow1** [reply]
https://www.notion.vip/wp-content/uploads/notion_vip-insert_toggle.png

**Reply by introt** [reply]
Here's a pure HTML + CSS demo. "Toggle lists" can help make notes really tidy by keeping the clutter (less-important details etc) at minimum, allowing one to glance over the most important points quickly.

![image](https://user-images.githubusercontent.com/26069787/143684140-c19ed207-e95e-4fe1-bc81-6be23a1f296e.png)
_Mixed list with the top-level togglable open._

![image](https://user-images.githubusercontent.com/26069787/143684154-eb52f16a-79d9-4434-8efa-afe2c94dc52f.png)
_Same list with both togglables open._

The HTML is really simple compared to how regular (sub)lists are handled; just replace `li` with `details` and wrap the bullet line in a `summary` tag.

```html
<ul>
  <li>Here's a regular old list element
    <ul>
      <li>Boring sublist item</li>
      <li>Another one, cluttering the view</li>
    </ul>
  </li>
  <details><summary>Here's a poor-man's toggleable list</summary>
    <ul>
      <details><summary>Using pure HTML</summary>
        <ul>
          <li>No need for JavaScript!</li>
          <li>Needs some CSS to fit in with the rest of the list...</li>
        </ul>
      </details>
      <li>Not supported in eg. MarkDown, but can be represented as a regular list there</li>
    </ul>
  </details>
</ul>
```

The minimal CSS needed for the togglables to fit in as shown above:

```css
ul { list-style-position: inside }
```

**Reply by introt** [reply]
Here's a quick'n'dirty implementation which replaces all lists: https://github.com/introt/zim-desktop-wiki/commit/d631b12b1d4eee3567fbe4e221675fd588f8939c

Playing around with this, I instantly fell in love with togglable index entries, turning the massive passive list into an interactive TreeView kinda thing. A proper implementation would complicate the html dumper somewhat, but doesn't seem too bad.

**Comment by introt** [comment]
About them toggle lists, here's togglable headers with pure HTML in two styles:

* "Recursive" (lower headers reside inside upper ones): [demo](https://introt.github.io/zim-desktop-wiki/data/templates/Folding_template_demo.html) | [template](https://raw.githubusercontent.com/introt/zim-desktop-wiki/sphinx/data/templates/Folding_outline.html)
* "Outline" (headers visible at all times): [demo](https://introt.github.io/zim-desktop-wiki/data/templates/Folding_outline_demo.html) | [template](https://raw.githubusercontent.com/introt/zim-desktop-wiki/sphinx/data/templates/Folding_with_index.html)

A few notes:

* Needs more CSS; I'm not a web dev, don't judge
* You can change the `<details>` tag to `<details open="">` to have the folds open by default.
    * A few lines of JavaScript could be added to open/close all folds (of the same level; take a look at how the summary styles are implemented to see how they could be identified)
* The implementation should be straight forward to port to other templates

Edit: don't know how I missed this, but there's literally a link to a "pure HTML+CSS" (no JS) TreeView on the Wikipedia page of tree views: https://experiments.wemakesites.net/css3-treeview.html - will look into that for the lists

Edit 2: added indentation to templates, making "Outline" into a "real" outline

Edit 3: related: #474

**Reply by introt** [reply]
@jaap-karssenberg I think I finally figured out "toggle lists (and headers)", they can be used to represent linear pages or lists as hierarchical outlines. As per Wikipedia: "Extended tree view is the central component of outliner applications, where each node consists of editable text."

While Zim is not an [outliner](https://en.wikipedia.org/wiki/Outliner) per se, toggleable headers and/or lists would also be useful for eg. not needing to jump between the text and a ToC, fitting different parts of a long page in a single window, active recall learning and getting a quick overview of a multi-level bullet list.

[Earlier](https://github.com/zim-desktop-wiki/zim-desktop-wiki/discussions/1667#discussioncomment-1446763) I mentioned waiting for embedded page views, but they're hardly necessary for reaping the mentioned benefits – if either is possible without those, I'd implement it as a plugin just to catch up with eg. Notion in this regard – though being able to integrate parts of other pages would certainly be a giant leap towards turning Zim into an outliner.

**Reply by introt** [reply]
![image](https://user-images.githubusercontent.com/26069787/144720176-29f8d812-6f81-4f26-94ff-e0dfce0df3f8.png)

https://en.wikipedia.org/wiki/Tree_structure#Outlines_and_tree_views

**Comment by jaap-karssenberg** [comment]
You mean you figured out how to do fixing in a GTK text view, or you mean
you figured out the use case?

Op za 4 dec. 2021 19:02 schreef introt ***@***.***>:

**Reply by introt** [reply]
Just the use case, sorry for the confusion!

**Comment by jaap-karssenberg** [comment]
Use case is clear for me. I was using outliner mode in vim before I started
Zim :)

Implantation wise I think it is possible to add folding to the page view.
There is an API to put icons in the margin. And there is hidden text
ranges. Just never took the time to really dive in


Op zo 5 dec. 2021 14:00 schreef introt ***@***.***>:

**Reply by introt** [reply]
For some reason my mind didn't connect "toggle lists" to outlining until seeing Notion in action (while myself opening more windows). Good to see we're _on the same outline_ ;)

**Comment by ghost** [comment]
as I have no skills regarding html or development, I really appreciate u guys talking about that topic!
I think toggle lists are one of the features that leads especially educational-users to switching to notion..at least in a few cases..
If that would be implemented one day I would be very happy, if not, I can still understand and appreciate zim wiki as it is.
thank u!

**Reply by introt** [reply]
Wish I had actual skills, just messing around lol (you're welcome) :)
The way Notion does images (fit page/cell width while instantly transiently zoomable) is also really great, would love it esp. for graphs and figures! Real shame it's such a closed system.

**Comment by brainbow1** [comment]
just wanted to ask if there is any chance that toggle lists will be implemented in the far future? I am no dev so Im not able to follow all of the conversation here in detail :-) would be happy about ur feedback :)

**Reply by introt** [reply]
<details><summary>The good news</summary>
There's a greater-than-zero chance that they will be implemented: they should be technically possible with our toolkit, and nobody is against adding them.</details>

<details><summary>The bad news</summary>It's been a year and as far as I know no one is actively working on them right now. Sorry.</details>

**Comment by brainbow1** [comment]
Okay, thanks a lot for ur answer :-) my hope will never die, I love zim anyway, thanks to all contributors ❤️
